File: python_solutions/day13.py
from enum import Enum


class Result(Enum):
    RIGHT = 0
    WRONG = 1
    TIE = 2


def compare(a, b):
    match (a, b):
        case ([a1, *a_rest], [b1, *b_rest]):
            if not (res := compare(a1, b1)) == Result.TIE:
                return res
            return compare(a_rest, b_rest)

        case ([], _):
            return Result.RIGHT

        case (_, []):
            return Result.WRONG

        case ([], []):
            return Result.TIE

        case (_, [_, *_]):
            return compare([a], b)

        case ([_, *_], _):
            return compare(a, [b])

        case _:
            if a < b:
                return Result.RIGHT
            elif a > b:
                return Result.WRONG
            else:
                return Result.TIE


def str_to_list(s: str) -> list[list | int]:
    if s == []:
        return s

    if s.startswith("["):
        s = s[1:]
    if s.endswith("]"):
        s = s[:-1]

    res = []

    i = 0
    while i < len(s):
        if s[i].isdigit():
            start = i
            i += 1
            while i < len(s) and s[i] != ",":
                i += 1
            res.append(int(s[start:i]))
            i += 1
            continue

        if s[i] == "[":
            paren_count = 1
            start = i
            i += 1
            while i < len(s) and paren_count > 0:
                if s[i] == "[":
                    paren_count += 1
                if s[i] == "]":
                    paren_count -= 1
                i += 1
            res.append(str_to_list(s[start:i]))

        i += 1

    return res


pairs = []
with open("inputs/day13.txt") as input:
    while line1 := input.readline().rstrip():
        line2 = input.readline().rstrip()
        pairs.append((str_to_list(line1), str_to_list(line2)))
        input.readline()


score = 0
for (i, (a, b)) in enumerate(pairs):
    if compare(a, b) == Result.RIGHT:
        score += i + 1

print(score)

File: src/day13.rs
use std::{cmp::Ordering, fs::File, io, io::BufRead};

#[derive(Debug, Clone)]
enum Packet {
    List(Vec<Box<Packet>>),
    Number(u32),
}

type PacketCouple = (Packet, Packet);

fn print_packet(packet: &Packet) {
    match packet {
        Packet::Number(n) => print!("{n}"),
        Packet::List(list) => {
            if list.len() == 0 {
                print!("[]");
                return;
            }
            print!("[");
            for (i, element) in list.iter().enumerate() {
                if i == list.len() - 1 {
                    print_packet(element);
                    print!("]");
                } else {
                    print_packet(element);
                    print!(",");
                }
            }
        }
    }
}

fn parse_packet(s: &str) -> Packet {
    if s == "[]" {
        return Packet::List(Vec::new());
    }

    if let Ok(n) = s.parse::<u32>() {
        return Packet::Number(n);
    }

    let mut chars = s.chars();
    let mut buffer = Vec::new();
    let mut elements = Vec::new();
    while let Some(c) = chars.next() {
        match c {
            '[' => {
                let mut paren_count = 1;
                while let Some(c) = chars.next() {
                    match c {
                        ']' => {
                            paren_count -= 1;
                            if paren_count > 0 {
                                buffer.push(c);
                            }
                        }
                        '[' => {
                            buffer.push(c);
                            paren_count += 1;
                        }
                        _ => buffer.push(c),
                    }
                    if paren_count == 0 {
                        break;
                    }
                }
                if buffer.len() == 0 {
                    elements.push(Packet::List(Vec::new()));
                } else {
                    elements.push(parse_packet(
                        buffer.into_iter().collect::<String>().as_str(),
                    ));
                    buffer = Vec::new();
                }
            }
            ',' => {
                if buffer.len() != 0 {
                    elements.push(parse_packet(
                        buffer.into_iter().collect::<String>().as_str(),
                    ));
                    buffer = Vec::new();
                }
            }
            _ => buffer.push(c),
        }
    }
    if buffer.len() != 0 {
        elements.push(parse_packet(
            buffer.into_iter().collect::<String>().as_str(),
        ));
    }

    Packet::List(elements.into_iter().map(Box::new).collect())
}

fn compare_packets((p1, p2): &PacketCouple) -> Option<bool> {
    match (p1, p2) {
        (Packet::Number(a), Packet::Number(b)) => match a.cmp(&b) {
            Ordering::Less => Some(true),
            Ordering::Greater => Some(false),
            Ordering::Equal => None,
        },
        (Packet::Number(a), Packet::List(list)) => compare_packets(&(
            Packet::List(vec![Box::new(Packet::Number(*a))]),
            Packet::List(list.to_vec()),
        )),
        (Packet::List(list), Packet::Number(b)) => compare_packets(&(
            Packet::List(list.to_vec()),
            Packet::List(vec![Box::new(Packet::Number(*b))]),
        )),
        (Packet::List(list1), Packet::List(list2)) => {
            let mut res = None;
            let (mut list1_iter, mut list2_iter) = (list1.iter(), list2.iter());
            while let None = res {
                res = match (list1_iter.next(), list2_iter.next()) {
                    (Some(a), Some(b)) => compare_packets(&(*a.clone(), *b.clone())),
                    (Some(_), None) => Some(false),
                    (None, Some(_)) => Some(true),
                    (None, None) => break,
                }
            }
            res
        }
    }
}

fn read_input(filepath: &str) -> Vec<PacketCouple> {
    let file = File::open(filepath).unwrap();
    let mut reader = io::BufReader::new(file).lines();
    let mut packets = Vec::new();

    while let Some(Ok(line1)) = reader.next() {
        let line2 = reader.next().unwrap().unwrap();

        packets.push((
            parse_packet(&line1.as_str()[1..(line1.len() - 1)]),
            parse_packet(&line2.as_str()[1..(line2.len() - 1)]),
        ));
        reader.next();
    }

    packets
}

pub fn part1() {
    let packets = read_input("inputs/day13.txt");

    let result = packets
        .into_iter()
        .enumerate()
        .filter_map(|(i, couple)| {
            if compare_packets(&couple).unwrap() {
                Some(i + 1)
            } else {
                None
            }
        })
        .fold(0, |a, b| a + b);

    println!("{}", result);
}
